src/v/raft/persisted_stm.cc

```diff
@@ -119,24 +120,40 @@ file_backed_stm_snapshot::load_snapshot() {
     }
 
     storage::snapshot_reader& reader = *maybe_reader;
-    iobuf meta_buf = co_await reader.read_metadata();
-    iobuf_parser meta_parser(std::move(meta_buf));
-    auto header = read_snapshot_header(meta_parser, _ntp, name());
-    if (!header) {
-        vlog(_log.warn, "Skipping snapshot {} due to old format", store_path());
-
-        // can't load old format of the snapshot, since snapshot is missing
-        // it will be reconstructed by replaying the log
-        co_await reader.close();
-        co_return std::nullopt;
-    }
     stm_snapshot snapshot;
-    snapshot.header = *header;
-    snapshot.data = co_await read_iobuf_exactly(
-      reader.input(), snapshot.header.snapshot_size);
+    std::exception_ptr ex{nullptr};
+    try {
+        iobuf meta_buf = co_await reader.read_metadata();
+        iobuf_parser meta_parser(std::move(meta_buf));
+        auto header = read_snapshot_header(meta_parser, _ntp, name());
+        if (!header) {
+            vlog(
+              _log.warn,
+              "Skipping snapshot {} due to old format",
+              store_path());
 
-    _snapshot_size = co_await reader.get_snapshot_size();
+            // can't load old format of the snapshot, since snapshot is missing
+            // it will be reconstructed by replaying the log
+            co_await reader.close();
+            co_return std::nullopt;
+        }
+        snapshot.header = *header;
+        snapshot.data = co_await read_iobuf_exactly(
+          reader.input(), snapshot.header.snapshot_size);
+
+        _snapshot_size = co_await reader.get_snapshot_size();
+    } catch (...) {
```

Comment:
Yeah exactly, this is the only way I know to do it without introducing a new function. Definitely kind of ugly, though I can't see how to wrap it up nicely either. Anyway, a problem for another day.